pkg/docker/docker.go

```diff
@@ -768,19 +892,21 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) error {
 		os.Exit(2)
 	}
 	return interrupt.New(dumpStack, removeContainer).Run(func() error {
-		// Attach to the container.
+		var attachError error
+		// Attach to the container on go thread (different than with fsouza, since it provided a non-blocking attach which we don't seem to have with k8s/engine-api)
+		// still preserve the flow of attaching before starting to handle various timing issues encountered in the past, as well as allow for --run option
 		glog.V(2).Infof("Attaching to container %q ...", containerName)
-		attachOpts := opts.asDockerAttachToContainerOptions()
-		attachOpts.Container = container.ID
-		if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
-			glog.V(0).Infof("error: Unable to attach to container %q with options %+v: %v", containerName, attachOpts, err)
-			return err
-		}
+		go func() {
+			attachOpts := opts.asDockerAttachToContainerOptions()
+			if attachError = d.kubeDockerClient.AttachToContainer(container.ID, attachOpts, opts.asDockerAttachToStreamOptions()); attachError != nil {
+				glog.V(0).Infof("error: Unable to attach to container %q with options %+v: %v", containerName, attachOpts, attachError)
+			}
+		}()
```

Comment:
in fact, the util.TimeoutAfter() stuff I mentioned earlier seems to do the exact trick with the channels.... I'm just reusing that.